### PR TITLE
EFM Integration Test Part 2

### PR DESCRIPTION
**Pull request by kc1116**
This PR updates the current EFM integration test and ensures that the network can successfully transition into recovery epoch after processing recover epoch governance transaction.

ref: https://github.com/onflow/flow-go/issues/6164

### CODE REVIEW

**Review by jordanschalm — COMMENTED** [review]
The test itself looks good, but I think we need to move it or adjust some configs to make sure it will run when we want 👇.

The integration tests are run in CI jobs grouped by package. The new epoch recovery test is in a new package `integration/tests/epochs/recover_epoch`, but there is no corresponding CI configuration, so I don't think the new test will actually run in CI, or when we run eg. `make epochs-tests`.

The easiest fix is to just add the new test to an existing package (cohort1 or cohort2). 

If it needs to be in a new package, then:
- we'll need to configure the new package to run in CI etc.
  - Need to add an entry to the Makefile [here](https://github.com/onflow/flow-go/blob/01fcf31d1eb7952c0e0276d8c386c5a5d57dd715/integration/Makefile#L48-L60)
  - And a new CI job [here](https://github.com/onflow/flow-go/blob/3cccef6d143f7a45e1540dd4c9efcdd8cc7507dd/.github/workflows/ci.yml#L320-L324)
- let's call the new package `cohort3` to keep naming consistent

File: fvm/environment/event_emitter.go [hunk 1]
```diff
@@ -201,8 +207,12 @@ func (emitter *eventEmitter) EmitEvent(event cadence.Event) error {
 
 		// skip limit if payer is service account
 		// TODO skip only limit-related errors
-		if !isServiceAccount && eventEmitError != nil {
-			return eventEmitError
+		if eventEmitError != nil {
```
Comment (durkmurder) [inline fvm/environment/event_emitter.go:210]: added some observability, previously error was swallowed.

File: engine/consensus/dkg/reactor_engine.go [hunk 2]
```diff
@@ -423,7 +423,7 @@ func (e *ReactorEngine) end(nextEpochCounter uint64) func() error {
 		if crypto.IsDKGFailureError(err) {
 			e.log.Warn().Err(err).Msgf("node %s with index %d failed DKG locally", e.me.NodeID(), e.controller.GetIndex())
 			err := e.dkgState.SetDKGEndState(nextEpochCounter, flow.DKGEndStateDKGFailure)
-			if err != nil {
+			if err != nil && !errors.Is(err, storage.ErrAlreadyExists) {
```
Comment (durkmurder) [inline engine/consensus/dkg/reactor_engine.go:426]: added just in case handling of sentinel error if recovery happened before failing the DKG

File: integration/tests/epochs/cohort2/epoch_recover_from_efm_test.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+	s.AwaitFinalizedView(s.Ctx, s.GetDKGEndView(), 2*time.Minute, 500*time.Millisecond)
+	// start the paused collection node now that we are in EFM
+	require.NoError(s.T(), ln.Start())
+
+	// get final view form the latest snapshot
+	epoch1FinalView, err := s.Net.BootstrapSnapshot.Epochs().Current().FinalView()
+	require.NoError(s.T(), err)
+
+	// wait for at least the first block of the next epoch to be sealed so that we can
+	// ensure that we are still in the same epoch after the final view of that epoch indicating we are in EFM
```
Comment (jordanschalm) [inline integration/tests/epochs/cohort2/epoch_recover_from_efm_test.go:140]:
```suggestion
	// Wait for at least the first view past the current epoch's original FinalView to be finalized.
	// At this point we can observe that an extension has been added to the current epoch, indicating EFM.
```

File: integration/tests/epochs/cohort2/epoch_recover_from_efm_test.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+	require.NoError(s.T(), err)
+	s.Client.Account().Keys[0].SequenceNumber++
+	require.NoError(s.T(), result.Error)
+
+	return result
+}
+
+// TestRecoverEpoch ensures that the recover epoch governance transaction flow works as expected and a network that
+// enters Epoch Fallback Mode can successfully recover. This test will do the following:
+// 1. Triggers EFM by turning off the sole collection node before the end of the DKG forcing the DKG to fail.
```
Comment (jordanschalm) [inline integration/tests/epochs/cohort2/epoch_recover_from_efm_test.go:119]: We have this comment and above we have:
> 	// to manually trigger EFM we assign very short dkg phase len ensuring the dkg will fail

I remember Khalil was trying a few different approaches to reliably trigger EFM, but I'm not sure which one ended up working. We should figure out which of these (Collection Node pause or short DKG phases) is actually causing EFM, and remove the comment and logic for the one that isn't.

Comment (durkmurder) [reply integration/tests/epochs/cohort2/epoch_recover_from_efm_test.go:119]: short DKG phase is somehow flaky, pausing collection node is very reliable though

File: module/dkg/recovery.go [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+		return fmt.Errorf("could not get epoch protocol state: %w", err)
+	}
+	currentEpochCounter := epochProtocolState.Epoch()
+
+	log := b.log.With().
+		Uint64("height", head.Height).
+		Uint64("view", head.View).
+		Uint64("epochCounter", currentEpochCounter).
+		Logger()
+	if epochProtocolState.EpochPhase() != flow.EpochPhaseCommitted {
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply module/dkg/recovery.go:101]: we are referring some block so it's hard to imagine a race condition, especially when we have started using reference block instead of finalized one.

File: cmd/bootstrap/run/epochs.go
Comment (AlexHentschel) [inline cmd/bootstrap/run/epochs.go:1]: ⚠️ 
I am a bit worried about the following line:
https://github.com/onflow/flow-go/blob/b9ab2da508aa895ed028059f7ac6a8665069d15c/cmd/bootstrap/run/epochs.go#L36
Actually, I am worried more generally about the filters (file `model/flow/filter/identity.go`).

`filter.IsValidProtocolParticipant` includes _all_ nodes that are joining or leaving. They still have an initial weight, but their `EpochParticipationStatus` is different than `EpochParticipationStatusActive`. These are _more nodes_ than the active participants of the _current_ epoch. What to do with joining and leaving nodes should be carefully considered. 

Furthermore, could you take a look at the `IsConsensusCommitteeMember` filter. I feel this filter may have the same problem that it includes joining or leaving nodes, while I think we intend it to only contain active nodes.

Comment (AlexHentschel) [reply cmd/bootstrap/run/epochs.go:1]: Update: I checked the [`IsConsensusCommitteeMember` filter](https://github.com/onflow/flow-go/blob/b9ab2da508aa895ed028059f7ac6a8665069d15c/model/flow/filter/identity.go#L127) and my concerns were unwarranted. For this filter, it is all good. 

The only aspect that I noticed is around documentation / naming: 
https://github.com/onflow/flow-go/blob/b9ab2da508aa895ed028059f7ac6a8665069d15c/model/flow/filter/identity.go#L144-L147

On the one hand, the `IsValidDKGParticipant` filter is only used in tests. On the other hand, it indicates that we can filter for DKG participants explicitly, which I think we can't do accurately anymore. 

**Suggestion:** remove  `filter.IsValidDKGParticipant` and remove all occurrences with `IsConsensusCommitteeMember`.

Comment (durkmurder) [reply cmd/bootstrap/run/epochs.go:1]: [`f832e98` (#6424)](https://github.com/onflow/flow-go/pull/6424/commits/f832e9802d2bc402cc8cfba58c47529559ef879a)

File: cmd/bootstrap/run/epochs.go [hunk 6]
```diff
@@ -135,6 +147,8 @@ func GenerateRecoverEpochTxArgs(log zerolog.Logger,
 	}
 
 	args := []cadence.Value{
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply cmd/bootstrap/run/epochs.go:149]: The target end time and duration values [are for the recovery epoch](https://github.com/onflow/flow-core-contracts/blob/feature/efm-recovery/contracts/epochs/FlowEpoch.cdc#L642-L643), _not_ the current epoch. 

Good catch. We need to change this. I would suggest the following:
- Keep the single duration flag, but rename it to `epoch-timing-recovery-duration`. Since recovery epochs may be unusual lengths, using a different name discourages re-using a flag value from the bootstrapping logic, which has a similar flag.
- Compute TargetEndTime as `currentEpoch.TargetEndTime() + timingDurationFlag`

File: module/dkg/recovery.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+	}
+	if safe {
+		log.Info().Msg("my beacon private key for the next epoch is safe, nothing to do")
+		return nil
+	}
+
+	myBeaconPrivateKey, safe, err := b.localDKGState.RetrieveMyBeaconPrivateKey(currentEpochCounter)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			log.Warn().Str(logging.KeySuspicious, "true").Msgf("no my beacon key for the current epoch has been found")
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply module/dkg/recovery.go:122]: I agree with not using `KeySuspicious` here, and with extending its documentation -- I suspect it is currently being used overly broadly because of the vague documentation ([example](https://github.com/onflow/flow-go/blob/01fcf31d1eb7952c0e0276d8c386c5a5d57dd715/network/p2p/builder/gossipsub/gossipSubBuilder.go#L69)). Adding a logging key for configuration problems makes sense to me. 

(By the way, there is also `KeyProtocolViolation` for _certain_ Byzantine behaviour.)

File: module/dkg/recovery.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("could not recover my beacon private key when initializing: %w", err)
+	}
+
+	return recovery, nil
+}
+
+// EpochFallbackModeExited implements handler from protocol.Consumer to perform recovery of the beacon private key when
+// this node has exited the epoch fallback mode.
+func (b *BeaconKeyRecovery) EpochFallbackModeExited(epochCounter uint64, _ *flow.Header) {
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply module/dkg/recovery.go]: > reasoning about correctness much simpler (no blocks can be finalized between the EFM exited event being emitted, and this component completing all its work)

I think that is a benefit short-term and tech-debt long-term: 
* Agree that is saves code and complexity now. However, now we are also fully immersed in the topic and arguably best equipped to pursue a more general implementation. 
* In the long term, we have made the following experience repeatedly:
   * We shortcut the more general implementation and correctness argument in the short term. This results in an edge-case that needs to be considered during future work. 
   * When we work on the framework in the future, we have forgotten about the edge case and accidentally break it. 
  
⚠️ Possible scenario here:
* In my mind, the notification design should _only_ guarantees causality (causality here means that the notifications are emitted strictly after the database operations and  order-preserving delivery of events to consumers). It does _not_ guarantee synchronicity within the consensus EventLoop. 
* The current implementation emits events synchronously within the consensus EventLoop. But that is an implementation detail in my opinion. We highlight it as an explanation why processing should be asynchronous, but in my opinion, synchronous event delivery should _not_ be a guaranteed property of the design for the following reason:
   * It is generally possible that some notifications trigger the emissions of other notifications. Unless all processing logic is implemented synchronously (which it isn't for performance reasons) only causality is maintained but not synchronicity.   
* Assume we are finding in the future that synchronously event-emission is notably slowing down consensus. According to the framework specification, it would be totally fine to spin out the event emission to a separate go routine (as it maintains causality). So in this particular case, we might make change the implementation such that the Event Loop channels the events to a separate go routine (after the db operations concluded), and this separate go routine serves the consumer while the EventLoop itself continues processing blocks. This fully preserves causality, but not synchronicity ❗  
  * We have learned in the past that it is a huge lift to go through all implementations and check for assumptions that we potentially could break. In practise we don't do that. 

Long story short:
* I am happy with using the caller's thread to process the notification (this only slows down the HotStuff EventLoop, which is fine given that this code path is used so rarely).
* I would prefer that we do _not_ assume that the source always delivers the notification synchronously on the EventLoop. This would limit massively the flexibility of the framework and generally does not hold in general (`protocol.Consumer` being an exception).

File: module/dkg/recovery.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply module/dkg/recovery.go:101]: possibly related: [my comment above](https://github.com/onflow/flow-go/pull/6424#discussion_r1841253717) (but not entirely sure)

File: module/dkg/recovery.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply module/dkg/recovery.go]: ### ⚠️ Technically, this consumer needs to be prioritized above other consumers

I think that `EpochFallbackModeExited` conceptually implies that the next epoch is ready. But `EpochFallbackModeExited` could be consumed by multiple consumers and we do not guarantee any order. Furthermore, we have a no clear ordering-convention with `EpochCommittedPhaseStarted`. 

It worried me, because I feel we are writing important state information in response to notifications or asynchronously, which should conceptually already be persisted by the time the notifications is emitted.

I think this is important to get right, because bugs / broken assumptions will be super hard to reproduce and debug.

File: module/dkg/recovery.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (durkmurder) [reply module/dkg/recovery.go]: [`b334c81` (#6424)](https://github.com/onflow/flow-go/pull/6424/commits/b334c811f5c88941e7a1fa3a7e067c4a5e291350)

File: cmd/bootstrap/run/epochs.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (durkmurder) [reply cmd/bootstrap/run/epochs.go:158]: [`2168811` (#6424)](https://github.com/onflow/flow-go/pull/6424/commits/2168811874396ea8651765cb1f2adcbe6f86032c)

File: cmd/bootstrap/run/epochs.go
Comment (durkmurder) [reply cmd/bootstrap/run/epochs.go:1]: [`36a3958` (#6424)](https://github.com/onflow/flow-go/pull/6424/commits/36a39583d42196113a3bb20b357cb377730a4764)

File: cmd/bootstrap/run/epochs.go [hunk 9]
```diff
@@ -69,44 +71,55 @@ func GenerateRecoverEpochTxArgs(log zerolog.Logger,
...
+	// Context: recovering from Epoch Fallback Mode requires that a sufficiency large fraction of consensus participants
+	// has valid random beacon keys (threshold signature scheme). The specific origin of those threshold keys is largely
+	// irrelevant. Running a centralized key generation process, using keys from an off-chain DKG, or reusing the random
+	// beacon keys from a prior epoch are all conceptually possible - provided the intersection between the consensus
+	// committee and the random beacon committee is large enough (for liveness).
+	// Implemented here:
+	// In a nutshell, we are carrying the current consensus and collector nodes forward into the next epoch (the Recovery
+	// Epoch). Removing or adding a small number of nodes here would be possible, but is not implemented at the moment.
+	// In all cases, a core requirement for liveness is: the fraction of consensus participants in the recovery epoch with
+	// valid random beacon should ber significantly larger than the threshold of the threshold-cryptography scheme.
```
Comment (AlexHentschel) [inline cmd/bootstrap/run/epochs.go:107]:
```suggestion
	// valid random beacon key should be significantly larger than the threshold of the threshold-cryptography scheme.
```

File: cmd/bootstrap/run/epochs.go [hunk 10]
```diff
@@ -69,44 +71,55 @@ func GenerateRecoverEpochTxArgs(log zerolog.Logger,
...
+	// The EFM Recovery State Machine will heuristically reject recovery attempts (specifically reject EpochRecover Service
+	// events) where when the intersection between consensus and random beacon committees is too small.
```
Comment (AlexHentschel) [inline cmd/bootstrap/run/epochs.go:109]:
```suggestion
	// events, when the intersection between consensus and random beacon committees is too small.
```

File: cmd/bootstrap/run/epochs.go [hunk 11]
```diff
@@ -135,14 +157,16 @@ func GenerateRecoverEpochTxArgs(log zerolog.Logger,
...
 		// staking phase end view
 		cadence.NewUInt64(currEpochFinalView + numViewsInStakingAuction),
 		// epoch end view
 		cadence.NewUInt64(currEpochFinalView + numViewsInEpoch),
-		// target duration
-		cadence.NewUInt64(targetDuration),
+		// recovery epoch target duration
+		cadence.NewUInt64(currEpochTargetEndTime + recoveryEpochTargetDuration),
 		// target end time
 		cadence.NewUInt64(currEpochTargetEndTime),
```
Comment (AlexHentschel) [inline cmd/bootstrap/run/epochs.go:171]: ⚠️ ❓ 🤯 

* Something is still wrong here I think. 
* Could you add a reference from where we are getting the transaction arguments from (if possible)? If we mix them up in order here, that will not be good.

File: cmd/util/cmd/epochs/cmd/recover_test.go [hunk 12]
```diff
@@ -77,30 +75,34 @@ func TestRecoverEpochHappyPath(t *testing.T) {
...
+		require.Equal(t, cadence.NewUInt64(finalView+flagNumViewsInStakingAuction), decodedValues[2])
 		// epoch end view
-		require.Equal(t, decodedValues[2], cadence.NewUInt64(finalView+flagNumViewsInEpoch))
+		require.Equal(t, cadence.NewUInt64(finalView+flagNumViewsInEpoch), decodedValues[3])
 		// target duration
-		require.Equal(t, decodedValues[3], cadence.NewUInt64(flagTargetDuration))
+		require.Equal(t, cadence.NewUInt64(targetEndTime+flagRecoveryEpochTargetDuration), decodedValues[4])
 		// target end time
 		expectedTargetEndTime, err := rootSnapshot.Epochs().Current().TargetEndTime()
 		require.NoError(t, err)
```
Comment (AlexHentschel) [inline cmd/util/cmd/epochs/cmd/recover_test.go:93]: same calculation as in the implementation, which I think is bugged

File: cmd/bootstrap/run/epochs.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply cmd/bootstrap/run/epochs.go:159]: The implementation, but it still seems to be inconsistent to me: see by comments below 
👉 https://github.com/onflow/flow-go/pull/6424#discussion_r1843137480
👉 https://github.com/onflow/flow-go/pull/6424#discussion_r1843142014

File: model/flow/dkg.go [hunk 13]
```diff
@@ -14,6 +14,9 @@ const (
 	DKGEndStateNoKey
 	// DKGEndStateDKGFailure - the underlying DKG library reported an error.
 	DKGEndStateDKGFailure
+	// DKGEndStateRecovered - this node has recovered its beacon key from a previous epoch.
+	// This occurs only for epochs which are entered through the EFM Recovery process (`flow.EpochRecover` service event).
+	DKGEndStateRecovered
```
Comment (AlexHentschel) [inline model/flow/dkg.go:19]: I find this conceptually confusing, because we are not recovering the DKG. Our goal is to arrive with a valid private threshold key (aka, the Random Beacon key) -- the DKG is only the means to get to this key on the happy path. 

```suggestion
	// RandomBeaconKeyRecovered - this node has recovered its beacon key from a previous epoch.
	// This occurs only for epochs which are entered through the EFM Recovery process (`flow.EpochRecover` service event).
	RandomBeaconKeyRecovered
```

File: model/flow/dkg.go [hunk 14]
```diff
@@ -26,6 +29,8 @@ func (state DKGEndState) String() string {
 		return "DKGEndStateNoKey"
 	case DKGEndStateDKGFailure:
 		return "DKGEndStateDKGFailure"
+	case DKGEndStateRecovered:
+		return "DKGEndStateRecovered"
```
Comment (AlexHentschel) [inline model/flow/dkg.go:33]:
```suggestion
	case RandomBeaconKeyRecovered:
		return "RandomBeaconKeyRecovered"
```

File: module/dkg/recovery_test.go [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+	recovery, err := NewBeaconKeyRecovery(unittest.Logger(), s.local, s.state, s.dkgState)
+	require.ErrorIs(s.T(), err, exception)
+	require.Nil(s.T(), recovery)
+}
+
+// TestNewBeaconKeyRecovery_NextEpochRetrieveMyBeaconPrivateKeyException tests a scenario:
+// - node is in epoch committed phase
+// - exception is thrown when trying to check if there is a safe beacon key for the next epoch
+// This is an unexpected error and should be propagated to the caller.
+func (s *BeaconKeyRecoverySuite) TestNewBeaconKeyRecovery_NextEpochRetrieveMyBeaconPrivateKeyException() {
```
Comment (AlexHentschel) [inline module/dkg/recovery_test.go:133]: I would recommend to include another test like this one, with the only difference that https://github.com/onflow/flow-go/blob/eb968f879fa9085c798360f237597db944757147/module/dkg/recovery.go#L112 returns an `storage.ErrNotFound`. In this case, we would expect the code to proceed, right?

File: module/dkg/recovery_test.go [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+		dkgState.AssertNumberOfCalls(s.T(), "UpsertMyBeaconPrivateKey", 0)
+	})
+}
+
+// TestNewBeaconKeyRecovery_RecoverKey tests a scenario:
+// - node is in epoch committed phase
+// - node doesn't have a safe beacon key for the next epoch
+// - node has a safe beacon key for the current epoch
+// - node is part of the DKG for the next epoch
+// In case like this we need try recovering the key from the current epoch.
```
Comment (AlexHentschel) [inline module/dkg/recovery_test.go:299]:
```suggestion
// In this case, the implementation should successfully recover the Random Beacon Private key from the current epoch.
```

File: module/dkg/recovery_test.go [hunk 17]
```diff
@@ -0,0 +1,369 @@
...
+}
+
+// TestEpochFallbackModeExited tests a scenario:
+// - node starts in epoch fallback phase
+// - when creating NewBeaconKeyRecovery we shouldn't attempt to recover the key since the epoch phase is not committed.
+// - node leaves EFM and transitions to the epoch committed phase
+// - node doesn't have a safe beacon key for the next epoch
+// - node has a safe beacon key for the current epoch
+// - node is part of the DKG for the next epoch
+// In case like this we need try recovering the key from the current epoch.
```
Comment (AlexHentschel) [inline module/dkg/recovery_test.go:338]:
```suggestion
// In this case, the implementation should successfully recover the Random Beacon Private key from the current epoch.
```

File: module/dkg/recovery_test.go [hunk 18]
```diff
@@ -0,0 +1,369 @@
...
+		performTest(dkgState)
+	})
+	s.Run("key-for-next-epoch-is-not-safe", func() {
+		dkgState := mockstorage.NewEpochRecoveryMyBeaconKey(s.T())
+		dkgState.On("RetrieveMyBeaconPrivateKey", s.nextEpochCounter).Return(nil, false, nil).Once()
+		performTest(dkgState)
+	})
+}
+
+// TestEpochFallbackModeExited tests a scenario:
```
Comment (AlexHentschel) [inline module/dkg/recovery_test.go:331]:
```suggestion
// TestEpochFallbackModeExited confirms successful key recovery
// when the recovery process is triggered by observing a `EpochFallbackModeExited` notification: 
```

**Review by AlexHentschel — APPROVED** [review]
well done. Amazing tests - very thorough and very well documented 🏅

File: module/dkg/recovery.go
[the remark this answers is not in the document]
Comment (durkmurder) [reply module/dkg/recovery.go:1]: https://github.com/onflow/flow-go/issues/6725

File: cmd/bootstrap/cmd/dkg.go [hunk 19]
```diff
@@ -46,6 +47,11 @@ func runBeaconKG(nodes []model.NodeInfo) dkg.DKGData {
 		log.Info().Msgf("wrote file %s/%s", flagOutdir, fmt.Sprintf(model.PathRandomBeaconPriv, nodeID))
 	}
 
+	indexMap := make(flow.DKGIndexMap, len(pubKeyShares))
+	for i, node := range nodes {
+		indexMap[node.NodeID] = i
+	}
+
 	// write full DKG info that will be used to construct QC
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply cmd/bootstrap/cmd/dkg.go:54]: I will take care of this in separate PR: https://github.com/onflow/flow-go/issues/6726.

File: engine/consensus/dkg/reactor_engine.go
[the remark this answers is not in the document]
Comment (durkmurder) [reply engine/consensus/dkg/reactor_engine.go:1]: [`b2d26a7` (#6424)](https://github.com/onflow/flow-go/pull/6424/commits/b2d26a78889e7f8d580051c7905d02755262cf2a)

File: cmd/bootstrap/run/epochs.go (hunk 11 above)
Comment (durkmurder) [reply cmd/bootstrap/run/epochs.go]: [`27150c1` (#6424)](https://github.com/onflow/flow-go/pull/6424/commits/27150c1c2ab0fc6cf20490bf62a4f0778332681b)

File: cmd/util/cmd/epochs/cmd/recover_test.go (hunk 12 above)
Comment (durkmurder) [reply cmd/util/cmd/epochs/cmd/recover_test.go]: [`27150c1` (#6424)](https://github.com/onflow/flow-go/pull/6424/commits/27150c1c2ab0fc6cf20490bf62a4f0778332681b)

File: engine/consensus/dkg/reactor_engine.go [hunk 20]
```diff
@@ -434,7 +434,7 @@ func (e *ReactorEngine) end(nextEpochCounter uint64) func() error {
 		if privateShare != nil {
 			// we only store our key if one was computed
 			err = e.dkgState.InsertMyBeaconPrivateKey(nextEpochCounter, privateShare)
-			if err != nil {
+			if err != nil && !errors.Is(err, storage.ErrAlreadyExists) {
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply engine/consensus/dkg/reactor_engine.go]: [`27150c1` (#6424)](https://github.com/onflow/flow-go/pull/6424/commits/27150c1c2ab0fc6cf20490bf62a4f0778332681b)

File: cmd/bootstrap/run/epochs.go [hunk 21]
```diff
@@ -155,13 +203,13 @@ func GenerateRecoverEpochTxArgs(log zerolog.Logger,
 		cadence.NewUInt64(currEpochFinalView + numViewsInStakingAuction),
 		// epoch end view
 		cadence.NewUInt64(currEpochFinalView + numViewsInEpoch),
-		// target duration
-		cadence.NewUInt64(targetDuration),
-		// target end time
+		// recovery epoch target duration
 		cadence.NewUInt64(currEpochTargetEndTime),
```
Comment (jordanschalm) [inline cmd/bootstrap/run/epochs.go:207]:
```suggestion
		cadence.NewUInt64(recoveryEpochTargetDuration),
```

File: cmd/util/cmd/epochs/cmd/recover_test.go [hunk 22]
```diff
@@ -85,11 +87,9 @@ func TestRecoverEpochHappyPath(t *testing.T) {
 		// epoch end view
 		require.Equal(t, cadence.NewUInt64(finalView+flagNumViewsInEpoch), decodedValues[3])
 		// target duration
-		require.Equal(t, cadence.NewUInt64(flagTargetDuration), decodedValues[4])
+		require.Equal(t, cadence.NewUInt64(expectedTargetEndTime), decodedValues[4])
```
Comment (jordanschalm) [inline cmd/util/cmd/epochs/cmd/recover_test.go:90]:
```suggestion
		require.Equal(t, cadence.NewUInt64(flagRecoveryEpochTargetDuration), decodedValues[4])
```

File: cmd/util/cmd/epochs/cmd/recover_test.go [hunk 23]
```diff
@@ -75,6 +75,8 @@ func TestRecoverEpochHappyPath(t *testing.T) {
 		currEpoch := rootSnapshot.Epochs().Current()
 		finalView, err := currEpoch.FinalView()
 		require.NoError(t, err)
+		expectedTargetEndTime, err := currEpoch.TargetEndTime()
```
Comment (jordanschalm) [inline cmd/util/cmd/epochs/cmd/recover_test.go:78]:
```suggestion
		currEpochTargetEndTime, err := currEpoch.TargetEndTime()
```

File: model/flow/dkg.go [hunk 24]
```diff
@@ -14,9 +14,9 @@ const (
 	DKGEndStateNoKey
 	// DKGEndStateDKGFailure - the underlying DKG library reported an error.
 	DKGEndStateDKGFailure
-	// DKGEndStateRecovered - this node has recovered its beacon key from a previous epoch.
-	// This occurs only for epochs which are entered through the EFM Recovery process ([flow.EpochRecover] service event).
-	DKGEndStateRecovered
+	// RandomBeaconKeyRecovered - this node has recovered its beacon key from a previous epoch.
+	// This occurs only for epochs which are entered through the EFM Recovery process (`flow.EpochRecover` service event).
```
Comment (jordanschalm) [inline model/flow/dkg.go:18]: Minor point on the formatting:
- Wrapping code references in `[]` [formats it as a link to the type in Godoc](https://go.dev/doc/comment#doclinks)
- Wrapping code references in ``` `` ``` doesn't have any formatting support (Markdown-style code spans are unsupported).

File: storage/badger/dkg_state.go [hunk 25]
```diff
@@ -155,14 +156,14 @@ func (keys *SafeBeaconPrivateKeys) RetrieveMyBeaconPrivateKey(epochCounter uint6
...
-			// retrieve the key - any storage error (including not found) is an exception
+		if endState == flow.DKGEndStateSuccess || endState == flow.RandomBeaconKeyRecovered {
+			// retrieve the key - any storage error (including `storage.ErrNotFound`) is an exception
 			var encodableKey *encodable.RandomBeaconPrivKey
 			encodableKey, err = keys.state.retrieveKeyTx(epochCounter)(txn)
 			if err != nil {
 				key = nil
 				safe = false
-				return fmt.Errorf("[unexpected] could not retrieve beacon key for epoch %d with successful DKG: %v", epochCounter, err)
+				return irrecoverable.NewExceptionf("[unexpected] could not retrieve beacon key for epoch %d with successful DKG: %v", epochCounter, err)
```
Comment (jordanschalm) [inline storage/badger/dkg_state.go:166]:
```suggestion
				return irrecoverable.NewExceptionf("could not retrieve beacon key for epoch %d with successful DKG: %v", epochCounter, err)
```
using `NewExceptionf` will already put a scary prefix on the message - no need for 2.

**Review by jordanschalm — APPROVED** [review]

File: model/flow/dkg.go (hunk 24 above)
Comment (durkmurder) [reply model/flow/dkg.go:18]: Good to know, thanks. I will make a PR to `feature/efm-recovery` to fix any of those.
